Terminal.Gui/View/ViewKeyboard.cs

```diff
@@ -1,464 +1,869 @@
+	/// <returns><see langword="false"/> if the key stroke was not handled. <see langword="true"/> if no
+	/// other view should see it.</returns>
+	/// <remarks>
+	/// Not all terminals support key distinct down/up notifications, Applications should avoid
+	/// depending on distinct KeyUp events.
+	/// <para>
+	/// Overrides must call into the base and return <see langword="true"/> if the base returns <see langword="true"/>.
+	/// </para>
+	/// <para>
+	/// See <see href="../docs/keyboard.md">for an overview of Terminal.Gui keyboard APIs.</see>
+	/// </para>
+	/// </remarks>
+	public virtual bool OnKeyUp (KeyEventArgs keyEvent)
+	{
+		if (!Enabled) {
+			return false;
 		}
 
-		/// <summary>
-		/// Returns all commands that are supported by this <see cref="View"/>.
-		/// </summary>
-		/// <returns></returns>
-		public IEnumerable<Command> GetSupportedCommands ()
-		{
-			return CommandImplementations.Keys;
+		// fire event
+		KeyUp?.Invoke (this, keyEvent);
+		if (keyEvent.Handled) {
+			return true;
 		}
 
-		/// <summary>
-		/// Gets the key used by a command.
-		/// </summary>
-		/// <param name="command">The command to search.</param>
-		/// <returns>The <see cref="Key"/> used by a <see cref="Command"/></returns>
-		public Key GetKeyFromCommand (params Command [] command)
-		{
-			return KeyBindings.First (kb => kb.Value.SequenceEqual (command)).Key;
+		if (Focused?.OnKeyUp (keyEvent) == true) {
+			return true;
 		}
 
-		/// <inheritdoc/>
-		public override bool ProcessHotKey (KeyEvent keyEvent)
-		{
-			if (!Enabled) {
-				return false;
-			}
+		return false;
 
-			var args = new KeyEventEventArgs (keyEvent);
-			if (MostFocused?.Enabled == true) {
-				MostFocused?.KeyPressed?.Invoke (this, args);
-				if (args.Handled)
-					return true;
-			}
-			if (MostFocused?.Enabled == true && MostFocused?.ProcessKey (keyEvent) == true)
-				return true;
-			if (_subviews == null || _subviews.Count == 0)
-				return false;
+	}
 
-			foreach (var view in _subviews)
-				if (view.Enabled && view.ProcessHotKey (keyEvent))
-					return true;
-			return false;
+	/// <summary>
+	/// Invoked when a key is released. Set <see cref="KeyEventArgs.Handled"/> to true to stop the key up event from being processed by other views.
+	/// <remarks>
+	/// Not all terminals support key distinct down/up notifications, Applications should avoid
+	/// depending on distinct KeyDown and KeyUp events and instead should use <see cref="KeyDown"/>.
+	/// <para>
+	/// See <see href="../docs/keyboard.md">for an overview of Terminal.Gui keyboard APIs.</see>
+	/// </para>
+	/// </remarks>
+	/// </summary>
+	public event EventHandler<KeyEventArgs> KeyUp;
+
+	#endregion KeyUp Event
+
+	#endregion Low-level Key handling
+
+	#region Key Bindings
+
+	/// <summary>
+	/// Gets the key bindings for this view.
+	/// </summary>
+	private Dictionary<Key, (Command [] commands, CommandScope scope)> KeyBindings { get; set; } = new Dictionary<Key, (Command [], CommandScope)> ();
```

Comment:
Is it worth adding a class instead of `(Command [] commands, CommandScope scope)`?

For example `CommandCollection`

That would future proof things if it was decided to add some other elements to what is a command?

Especially for method like `GetKeyBindingWithScope` having a concrete type I think is better than this tuple.

Reply:
I concur. It was fun using the tuple, and i was surprised by how easy it was to modify the code like this. But it does seem fragile. I'll look into doing as you suggest.

Reply:
Due to this https://github.com/gui-cs/Terminal.Gui/pull/2927#issuecomment-1854220989, I think that class also can be contain a flag on the scope on which key bindings will handle (`OnKeyDown`, `OnKeyPress` or `OnKeyUp`). What do you think?